### ISSUE TITLE
When deployAtEnd is set, d-m-p should push docker images after a multi-module build

### ISSUE THREAD

**Issue by sparsick** [fabric8io/docker-maven-plugin#1266]
### Description
The idea is cloned from maven-deploy-plugin. When the user property deployAtEnd is set, d-m-p should push all Docker images at the end of multi module build. These docker images are created during this multi module build.

* If it's a feature request, what is your use case : In a CI build I want to ensure that the multi module build runs successfully before the docker images are pushed to the registry. 

*Workaround:*
```
mvn deploy -DdeployAtEnd -Ddocker.skip.push
mvn docker:push
```

**Comment by rhuss** [comment]
I see the use case, but how would a parent pom know which images to push ? 'would have to check how the deploy plugin deals with this and maybe one can also already achieve that with d-m-p with some clever lifecycle bindings, so that the sub-modules push e.g. on the `postInstall` phase (which is reached only when the build doesn't break before.

**Comment by sparsick** [comment]
I discussed the solution of the deploy plugin with @khmarbaise. He recommends me to use the [Maven deployer extension](https://github.com/khmarbaise/maven-deployer-extension) for this use case in deploy plugin. Maybe for this issue it would be a good idea to write a docker push extension. What do you think?

**Comment by rhuss** [comment]
I'm not very familiar with extensions, but sounds like a good idea ;-) Such an extension then could pick up all images to push at the end of the whole run.

I would happy to help to integrate any PR which goes in this direction.

**Comment by MariusSchmidt** [comment]
Yes, please please make it work, it is very much needed :). We often have this kind of structure

Parent-Aggregator
|-- Module A (e.g. a common Business  API)
|-- Module B, depends on A, defines image X
|-- Module C, depends on A, defines image Y
|-- Module D, product assembly, defines E2E tests and runs them utilizing images. Defines E2E test image

We usually only ever want to push images, after all E2E Tests were done.

**Comment by sparsick** [comment]
I started a [PoC](https://github.com/sparsick/maven-docker-push-extension) for this issue.

Currently, test infrastructure and the skeleton is set up.

**Comment by sparsick** [comment]
I release the first version of the Maven Extension. It is available in [Maven Central](https://search.maven.org/artifact/com.github.sparsick.maven.docker.extension/maven-docker-extension)

**Comment by tpischke-bedag** [comment]
This is a great idea, but I see that both this plugin and its inspiration plugin are in the POC Phase. Any plans to make either of these extensions production-ready, or push the change upstream to the modified plugins?

**Comment by sparsick** [comment]
@tpischke-bedag Could you test the current release version of the Maven extension and give me feedback if it works in your use case? IMHO there are some dependencies clean-ups to do. If it is better for the community, I can move the Maven extension to this project (@rhuss what do think?). But we cannot merge it with the plugin because a Maven extension is a separate concept towards a Maven plugin.

**Comment by TIBCOrkrajews** [comment]
Thus looks promising, but of course, there's always something new to deal with – in particular, some users are likely to move away from fabric8 and to https://www.eclipse.org/jkube/ , which is apparently its designated successor. But that can be dealt with later.

**Comment by rohanKanojia** [comment]
@TIBCOrkrajews : JKube is the successor to Fabric8 Maven Plugin not Docker Maven Plugin. This project is still alive and maintained.

**Comment by tpischke-bedag** [comment]
> @tpischke-bedag Could you test the current release version of the Maven extension and give me feedback if it works in your use case? IMHO there are some dependencies clean-ups to do. If it is better for the community, I can move the Maven extension to this project (@rhuss what do you think?). But we cannot merge it with the plugin because a Maven extension is a separate concept towards a Maven plugin.

We solved this by adding a dedicated deployment maven Module for the moment, but we will try testing this and providing feedback hopefully soon.

**Comment by tpischke-bedag** [comment]
> @tpischke-bedag Could you test the current release version of the Maven extension and give me feedback if it works in your use case? IMHO there are some dependencies clean-ups to do. If it is better for the community, I can move the Maven extension to this project (@rhuss what do you think?). But we cannot merge it with the plugin because a Maven extension is a separate concept towards a Maven plugin.

Just tried a quick test of the maven-docker-extension. Although the extension prevented the images from being pushed during each module's build, the images were also never pushed at the end of the build.
